### PR TITLE
HADOOP-16269. ABFS: add listFileStatus with StartFrom.

### PR DESCRIPTION
**Pull request by DadanielZ**
- Add support to list entries in a path from an entry name in lexical order(Azure Storage Service returns the entries in lexical order)
- This support is added to AzureBlobFileSystemStore and won't be exposed to FS level api.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 33 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1036 | trunk passed |
| +1 | compile | 33 | trunk passed |
| +1 | checkstyle | 22 | trunk passed |
| +1 | mvnsite | 34 | trunk passed |
| +1 | shadedclient | 670 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 38 | trunk passed |
| +1 | javadoc | 20 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 25 | the patch passed |
| +1 | compile | 23 | the patch passed |
| +1 | javac | 23 | the patch passed |
| -0 | checkstyle | 14 | hadoop-tools/hadoop-azure: The patch generated 4 new + 2 unchanged - 0 fixed = 6 total (was 2) |
| +1 | mvnsite | 26 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 675 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 46 | the patch passed |
| +1 | javadoc | 17 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 77 | hadoop-azure in the patch passed. |
| +1 | asflicense | 23 | The patch does not generate ASF License warnings. |
| | | 2901 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-768/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/768 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux e1f8804a2a8c 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a703dae |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/1/testReport/ |
| Max. process+thread count | 445 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 1]
```diff
@@ -583,6 +620,69 @@ public FileStatus getFileStatus(final Path path) throws IOException {
     return fileStatuses.toArray(new FileStatus[fileStatuses.size()]);
   }
 
+  // generate continuation token for xns account
+  private String generateContinuationTokenForXns(final String firstEntryName) {
+    if (firstEntryName == null || firstEntryName.isEmpty() || firstEntryName.startsWith(AbfsHttpConstants.ROOT_PATH)) {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: generally done in Preconditions.checkArgument() through the code these days

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 2]
```diff
@@ -583,6 +620,69 @@ public FileStatus getFileStatus(final Path path) throws IOException {
...
+    // Notice: non-xns continuation token requires full path (first "/" is not included) for startFrom
+    final String startFrom = (path.isEmpty() || path.equals(ROOT_PATH))
+            ? firstEntryName
+            : path + ROOT_PATH + firstEntryName;
+
+    SimpleDateFormat simpleDateFormat = new SimpleDateFormat(TOKEN_DATE_PATTERN, Locale.US);
+    String date = simpleDateFormat.format(new Date());
+
+    StringBuilder sb = new StringBuilder(path.length() + startFrom.length() + date.length() + 24);
+    sb.append(String.format("%06d", path.length())).append(CHAR_EXCLAMATION_POINT)
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: any reason not to use a single large String.format? I'm just wondering which version is the most clear as to what is being constructed. If the CHAR_EXCLAMATION_POINT was just "!" inlined the case for a single String.format would be clear. Here: not so sure

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/CRC64.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.fs.azurebfs.utils;
+
+/**
+ * CRC64 implementation for AzureBlobFileSystem.
+ */
+public class CRC64 {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/CRC64.java]: did you write this from scratch or take if from an existing implementation. If from scratch, we should think about a unit test or two. If from somewhere else, we'll need the provenance to be sure its OK to add in. If from colleagues, well, did they write any unit tests?

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  @Parameterized.Parameter(1)
+  public String startFrom;
+
+  @Parameterized.Parameter(2)
+  public int expectedStartIndexInArray;
+
+  @Parameterized.Parameter(3)
+  public boolean expectedResult;
+
+  @Parameterized.Parameters(name = "Testing path \"{0}\", startFrom: \"{1}\",  Expecting result : {3}") // Test path
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: I never knew about this! I shall annotate my parameterized tests this way too!

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+  // compare the file status
+  private void verifyFileStatus(FileStatus[] listResult, Path parentPath, int startIndexInSortedName) throws IOException {
+    if (startIndexInSortedName == -1) {
+      Assert.assertTrue("Expected empty FileStatus array", listResult.length == 0);
+      return;
+    }
+
+    FileStatus[] allFileStatuses = fs.listStatus(parentPath);
+    Assert.assertTrue("number of dir/file doesn't match",
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: use assertEquals()

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+  private void prepareTestFiles() throws IOException {
+    final AzureBlobFileSystem fs = getFileSystem();
+    // created 2 level file structures
+    for (String levelOneFolder : SORTED_ENTRY_NAMES) {
+      Path levelOnePath = new Path("/" + levelOneFolder);
+      Assert.assertTrue(fs.mkdirs(levelOnePath));
+      for (String fileName : SORTED_ENTRY_NAMES) {
+        Path filePath = new Path(levelOnePath, fileName);
+        fs.create(filePath);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: this doesn't close the stream. Should be `fs.create(path).close()`

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        Assert.assertTrue(fs.exists(filePath));
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: use ContractTestUtils.assertIsFile()

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 1 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: updated with Precondition.check

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/CRC64.java (hunk 3 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/CRC64.java]: Yes, this is written from scratch, also notice that ABFS crc64 use its own  polynomial. 
Unit test are added with hardcoded expected results, which were retrieved from azure service.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 2 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: you are correct. Included "!" in the string format.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java (hunk 4 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: haha, I find these annotation helpful to check on the failed tests.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java (hunk 5 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: updated

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java (hunk 7 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: updated with ContracTestUtils for file creation and verification.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java (hunk 6 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: updated with ContracTestUtils.touch

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 31 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1088 | trunk passed |
| +1 | compile | 27 | trunk passed |
| +1 | checkstyle | 18 | trunk passed |
| +1 | mvnsite | 32 | trunk passed |
| +1 | shadedclient | 694 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 40 | trunk passed |
| +1 | javadoc | 23 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 25 | the patch passed |
| +1 | compile | 24 | the patch passed |
| +1 | javac | 24 | the patch passed |
| -0 | checkstyle | 13 | hadoop-tools/hadoop-azure: The patch generated 4 new + 2 unchanged - 0 fixed = 6 total (was 2) |
| +1 | mvnsite | 28 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 746 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 49 | the patch passed |
| +1 | javadoc | 17 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 86 | hadoop-azure in the patch passed. |
| +1 | asflicense | 24 | The patch does not generate ASF License warnings. |
| | | 3045 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-768/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/768 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 9e551bc6d3e3 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3758270 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/2/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 527 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1145 | trunk passed |
| +1 | compile | 35 | trunk passed |
| +1 | checkstyle | 21 | trunk passed |
| +1 | mvnsite | 38 | trunk passed |
| +1 | shadedclient | 696 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 47 | trunk passed |
| +1 | javadoc | 21 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 27 | the patch passed |
| +1 | compile | 24 | the patch passed |
| +1 | javac | 24 | the patch passed |
| -0 | checkstyle | 14 | hadoop-tools/hadoop-azure: The patch generated 6 new + 2 unchanged - 0 fixed = 8 total (was 2) |
| +1 | mvnsite | 27 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| -1 | shadedclient | 109 | patch has errors when building and testing our client artifacts. |
| +1 | findbugs | 45 | the patch passed |
| +1 | javadoc | 17 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 80 | hadoop-azure in the patch passed. |
| +1 | asflicense | 23 | The patch does not generate ASF License warnings. |
| | | 2967 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-768/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/768 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 53320c43a750 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 3758270 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/3/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/3/testReport/ |
| Max. process+thread count | 411 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/3/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by DadanielZ** [comment]
Yetus is complaining about "shadedclient", but I tried the same command in my local env, there is no issue.
Any ideas?

**Comment by steveloughran** [comment]
No idea.
 @aw-was-here -your thoughts?

**Comment by DadanielZ** [comment]
Tested with my US-West account:
non-xns account:
Tests run: 40, Failures: 0, Errors: 0, Skipped: 0
Tests run: 342, Failures: 0, Errors: 0, Skipped: 207
Tests run: 190, Failures: 0, Errors: 0, Skipped: 15

xns-account:
Tests run: 40, Failures: 0, Errors: 0, Skipped: 0
Tests run: 342, Failures: 0, Errors: 0, Skipped: 21
Tests run: 190, Failures: 0, Errors: 0, Skipped: 15

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 30 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1171 | trunk passed |
| +1 | compile | 28 | trunk passed |
| +1 | checkstyle | 22 | trunk passed |
| +1 | mvnsite | 30 | trunk passed |
| +1 | shadedclient | 685 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 42 | trunk passed |
| +1 | javadoc | 20 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 28 | the patch passed |
| +1 | compile | 23 | the patch passed |
| +1 | javac | 23 | the patch passed |
| -0 | checkstyle | 15 | hadoop-tools/hadoop-azure: The patch generated 5 new + 2 unchanged - 0 fixed = 7 total (was 2) |
| +1 | mvnsite | 29 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 736 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 45 | the patch passed |
| +1 | javadoc | 18 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 80 | hadoop-azure in the patch passed. |
| +1 | asflicense | 26 | The patch does not generate ASF License warnings. |
| | | 3113 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-768/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/768 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 9265f7433f05 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / dead9b4 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/4/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/4/testReport/ |
| Max. process+thread count | 410 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/4/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 8]
```diff
@@ -522,15 +537,38 @@ public FileStatus getFileStatus(final Path path) throws IOException {
...
+  }
+
+  /**
+   * @param path Path the list path.
+   * @param startFrom the entry name that list results should start with.
+   *                  Notice that startFrom can be a non-existent entry, then the list request response contains
+   *                  all entries after this non-existent entry in lexical order.
+   * @return the entries in the path in lexical order from "startFrom".
+   * */
+  public FileStatus[] listStatus(final Path path, final String startFrom) throws IOException {
```
Comment (brkyvz) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: this doesn't need to be a public API

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java [hunk 9]
```diff
@@ -522,15 +537,38 @@ public FileStatus getFileStatus(final Path path) throws IOException {
...
+   * @return the entries in the path.
+   * */
   public FileStatus[] listStatus(final Path path) throws IOException {
-    LOG.debug("listStatus filesystem: {} path: {}",
+    return listStatus(path, null);
+  }
+
+  /**
+   * @param path Path the list path.
+   * @param startFrom the entry name that list results should start with.
```
Comment (brkyvz) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Could you give an example of `startFrom` please? From the tests it seems like the `startFrom` is supposed to be a relative name within `path`?
Should this API be called like:
```java
listStatus("/some-folder", "file-in-folder")
listStatus("/some-folder", "file-in")
```
or
```java
listStatus("/some-folder", "/some-folder/file-in-folder")
listStatus("/some-folder", "/some-folder/file-in")
```
from the test cases it seems like the former.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/TestAbfsCrc64.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.hadoop.fs.azurebfs.utils.CRC64;
+/**
+ * Test for Crc64 in AzureBlobFileSystem, notice that ABFS CRC64 has its own polynomial.
+ * */
+public class TestAbfsCrc64 {
+
+  @Test
+  public void tesCrc64Compute() {
+    CRC64 crc64 = new CRC64();
+    final String[] testStr = {"#$", "dir_2_ac83abee", "dir_42_976df1f5"};
```
Comment (ajfabbri) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/TestAbfsCrc64.java]: Shame we have to paste in a new CRC64 implementation but I didn't find others available in the codebase. Thanks for including a unit test.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java (hunk 4 above)
Comment (ajfabbri) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemStoreListStatusWithRange.java]: Yes, this is pretty nice.

**Comment by ajfabbri** [comment]
This looks pretty good overall. I think we need some documentation and proposed use case (who will call this method). Also a proposal how we integrate this with the greater FileSystem APIs would be good. We generally want to keep the FS implementations on the same API or they can diverge over time and create a mess. Depending on the greater API discussion, the listStatus method probably needs some Private Unstable annotations.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 9 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Yes, the former one is correct. `startFrom` is the name of an entry,  not a full path.
I will add an example here.

**Comment by DadanielZ** [comment]
@ajfabbri Because it relies on Azure service to return results in lexical order, so I am not sure if this need to be integrated to FileSystem API. I will add some private Unstable annotations for now.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java (hunk 8 above)
Comment (DadanielZ) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: From what I know hadoop-azure jar will be imported by customer code, so I keep it public here. I will add unstable annotation here for now.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 30 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1047 | trunk passed |
| +1 | compile | 27 | trunk passed |
| +1 | checkstyle | 18 | trunk passed |
| +1 | mvnsite | 34 | trunk passed |
| +1 | shadedclient | 177 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 24 | trunk passed |
| 0 | spotbugs | 54 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 52 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 25 | the patch passed |
| +1 | compile | 24 | the patch passed |
| +1 | javac | 24 | the patch passed |
| -0 | checkstyle | 15 | hadoop-tools/hadoop-azure: The patch generated 4 new + 2 unchanged - 0 fixed = 6 total (was 2) |
| +1 | mvnsite | 29 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 746 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 21 | the patch passed |
| +1 | findbugs | 55 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 81 | hadoop-azure in the patch passed. |
| +1 | asflicense | 31 | The patch does not generate ASF License warnings. |
| | | 2593 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-768/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/768 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 7924d0c9a0e0 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / d6b7609 |
| Default Java | 1.8.0_191 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/5/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/5/testReport/ |
| Max. process+thread count | 411 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-768/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by DadanielZ** [comment]
Can anyone help to review or commit this PR?

**Comment by steveloughran** [comment]
+1, committed.

thanks

I'm going to leave a warning note on the JIRA about how these internal things are brittle and may go away.

**Pull request closed without merging**